### ISSUE TITLE
Show which journalist sent a given reply

### ISSUE THREAD

**Issue by redshiftzero** [freedomofpress/securedrop-client#76]
Each conversation in the conversation view is between one source and multiple journalists. We are storing this information and should show on the conversation view which journalist sent each reply.

If a first name and last name are set, they should be used instead of a username. 

# Acceptance Criteria

**Given** that there is a conversation with at least one reply to a source
**When** I view that conversation in the SecureDrop client
**Then** individual replies to the source should be identified by the journalist's real name (if set) or user name (if not set).

**Given** that there is such a conversation
**When** I view it in the Source Interface
**Then** the author of the reply should _not_ be identified.

# User Story

As a journalist, I want to know which reply was sent by which journalist so that I can understand the conversation history. 

[See the [user stories spreadsheet](https://docs.google.com/spreadsheets/d/1DQEQp-0e6zN-pCPa0qqY-MGtmpc2SN5mzUsR4nEkpCQ/edit#gid=0) for current prioritization of this functionality.]

*Updated by @eloquence on 2019-06-10 to reflect first name / last name functionality and add acceptance criteria.*

**Comment by ninavizz** [comment]
Static wire demonstrating threading. Sources should NOT see the ID badges, but within the Client journalists should all see their own (upper-right, always; in convo thread when relevant) and their colleague's (in convo threads) ID badges. 

Letter in the badge is determined by the first letter of a username, or the person's first name (if/when that separate feature is added). https://drive.google.com/open?id=1GxXbYgQAR7nHlnvXxPFZ_oeA1NmZmf0e

Cheerio!

**Comment by heartsucker** [comment]
My interpretation of this is that we need a server-side API to pull the list of current journalists. This is a minimum requirement for this feature.

Additionally, this would allow journalists to update their display name (see https://github.com/freedomofpress/securedrop/issues/1592) and have that change propagated to the clients.

**Comment by eloquence** [comment]
I've updated the top level comment to reflect the recent addition of first name / last name functionality. Like #411, this is blocked on https://github.com/freedomofpress/securedrop-sdk/issues/87 (SDK support for first names / last names) unless we decide to do a first iteration with just usernames.

The current [design spec](https://app.zeplin.io/project/5c807ea562f734bd2756b243/screen/5cca0214ece43c348461f1d1) for the conversation view only reflects display of initials; if we ever want to show more than that (e.g., in a hover state), we'll need a design spec for that, as well.

**Comment by eloquence** [comment]
Still unclear whether this will make it to beta. Leaving a note here that I want to make sure we carefully reason through the behavior when user accounts are deleted on the server:

- How do we want to deal with loss of attribution information on the server?

- Are there any negative side effects to anticipate due to the use of both IDs and UUIDs on the server, e.g., could there be a situation where a new user account gets an ID of an old one, which would cause historical replies to be incorrectly attributed when they are newly downloaded?

**Comment by eloquence** [comment]
(Kept on beta but marked as stretch for now, more likely Post-Beta IMO)

**Comment by eloquence** [comment]
As expected, moved to post-beta.

**Comment by eloquence** [comment]
This is not quite ready for dev yet, but we've agreed to put our heads together (at least @creviera @eloquence @ninavizz) to finalize the design and implementation plan for a first iteration, as part of the 7/23-8/5 sprint.

**Comment by creviera** [comment]
update: working on a spike to see how this can be split into a couple or few prs, looks like we need to make a few backend changes to make sure a journalist for a DraftReply cannot be None and to pass around the logged in user when authentication changes (instead of just whether or not the user is authenticated) so that we can get the name of the sender of a reply. 

can't remember if we discussed hover-overs to show the fullname but that would be a useful thing to add

**Comment by creviera** [comment]
you can see how it looks like the journalist is `None` here when we are dealing with a `DraftReply`:
![Screenshot from 2020-07-31 16-46-29](https://user-images.githubusercontent.com/4522213/89088479-6d6ca280-d34d-11ea-8536-0fb08f76c396.png)

i'm at a good stopping point so i'll see if it's just something silly i was doing when i return

**Comment by creviera** [comment]
I created a separate issue for first handling backend changes: https://github.com/freedomofpress/securedrop-client/issues/1133

**Comment by eloquence** [comment]
We (Mickael/Allie/Nina/Erik) discussed the first iteration of this feature in some more detail today. We agreed that the first iteration will only use two colors, the `A`/`1` combination from https://scene.zeplin.io/project/5c807ea562f734bd2756b243/screen/5cba6b2f3367280c23df2be4, and a tooltip for showing the full name. The first iteration should use the `A` color to identify the currently logged in user (if any) both in the AuthWidget and for replies attributed to that user.

Once that's done, we can resolve _this_ issue, and track further UX changes (more colors, improved hover behavior) in separate tickets.

Still TBD: how to render deleted users in the first iteration.

**Comment by eloquence** [comment]
For the 8/20-9/2 sprint, we've committed to landing the backend dependencies for this change, including an SDK release (https://github.com/freedomofpress/securedrop-sdk/issues/127), and to get a draft PR for the reply badges functionality up which may not be fully ready for review yet (e.g., handling of deleted users may still be pending).

**Comment by creviera** [comment]
Going to make a draft PR of this branch https://github.com/freedomofpress/securedrop-client/tree/journalist-initials-in-reply until https://github.com/freedomofpress/securedrop-sdk/issues/128 is resolved

**Comment by eloquence** [comment]
(Removing epic from the current sprint; we _are_ tracking #1142 as part of the current sprint, which may fully resolve this issue as an MVP, or we may split off some follow-up work before we call it ready for release).

**Comment by ninavizz** [comment]
Woop Woop!

```
<svg width="36" height="34" viewBox="0 0 36 34" xmlns="http://www.w3.org/2000/svg"><style>.a{opacity:0.9;}</style><title>  Group 37</title><g fill="none"><g fill="#FFF"><path d="M16.7 3C10.6 9.9 9.7 9.7 6.7 1 9.8 9.7 9.2 10.4 0 8.5 9.2 10.4 9.4 11.2 3.3 18 9.4 11.1 10.2 11.3 13.3 20 10.2 11.3 10.8 10.6 20 12.5 10.8 10.7 10.4 9.9 16.7 3" class="a"/><path d="M36 16.6C26.5 22 24 21.5 18.6 12 24 21.5 23.5 24 14 29.4 23.5 24 26 24.5 31.4 34 26 24.5 26.5 22 36 16.6" class="a"/><path d="M23.5 0C24.8 3.5 24.5 4.2 21 5.5 24.5 4.2 25.2 4.5 26.5 8 25.2 4.5 25.5 3.8 29 2.5 25.5 3.8 24.8 3.5 23.5 0"/><path d="M7.5 22C8.8 25.5 8.5 26.2 5 27.5 8.5 26.2 9.2 26.5 10.5 30 9.2 26.5 9.5 25.8 13 24.5 9.6 25.8 8.9 25.5 7.5 22" class="a"/></g></g></svg>
```

**Comment by creviera** [comment]
## Update

### Before

Currently, on `main`, all journalist replies are the same color, except when they fail to send (then they are red) or fail to decrypt (then they are gray). Here is an example of the current client:

![current-without-journalist-badges](https://user-images.githubusercontent.com/4522213/95363345-32727800-0884-11eb-8abf-07aadc6cb83c.png)

### After

Here is an example of how [the PR that closes this issue](https://github.com/freedomofpress/securedrop-client/pull/1142) (ready for review when things settle down for folks and they have time) shows which journalist sent which message, whether the account exists or not (a deleted accounts show the sparkles/stars svg Nina provided here https://github.com/freedomofpress/securedrop-client/issues/76#issuecomment-703999320) and which replies came from the user who is currently logged in (purple badges):

![new-journalist-badges](https://user-images.githubusercontent.com/4522213/95363520-6a79bb00-0884-11eb-86dd-30f819479200.png)

### Followup

I opened a [followup discussion issue](#1154) for us to use as a place to capture ideas around how to display errors now that we have reply badges and use colors to mean more than just whether or not there was an error.

**Comment by ninavizz** [comment]
@creviera I polished the twinklies a bit so they're less-retro and more refined. More twinkly, less sputnik. It's subtle, and may not be worth the late inclusion into the merge, but FYI:

```
<svg width="41" height="36" viewBox="0 0 41 36" xmlns="http://www.w3.org/2000/svg"><style>.a{opacity:0.7;}</style><title>  Twinklies</title><g fill="none"><g fill="#FFF"><path d="M16.7 3C10.6 9.9 9.7 9.7 6.7 1 9.8 9.7 9.2 10.4 0 8.5 9.2 10.4 9.4 11.2 3.3 18 9.4 11.1 10.2 11.3 13.3 20 10.2 11.3 10.8 10.6 20 12.5 10.8 10.7 10.4 9.9 16.7 3" class="a"/><path d="M37.7 13.2C27.4 19.2 24.7 18.5 18.7 8.2 24.7 18.5 24 21.3 13.7 27.3 24 21.3 26.8 21.9 32.7 32.2 26.8 21.9 27.4 19.2 37.7 13.2" opacity="0.5"/><path d="M24.5 16C25.8 19.5 25.5 20.2 22 21.5 25.5 20.2 26.2 20.5 27.5 24 26.2 20.5 26.5 19.8 30 18.5 26.6 19.8 25.9 19.5 24.5 16"/><path d="M23.5 0C24.8 3.5 24.5 4.2 21 5.5 24.5 4.2 25.2 4.5 26.5 8 25.2 4.5 25.5 3.8 29 2.5 25.5 3.8 24.8 3.5 23.5 0" class="a"/><path d="M8.5 7C9.8 10.5 9.5 11.2 6 12.5 9.5 11.2 10.2 11.5 11.5 15 10.2 11.5 10.5 10.8 14 9.5 10.5 10.8 9.8 10.5 8.5 7" opacity="0.8"/><path d="M11.5 24C12.8 27.5 12.5 28.2 9 29.5 12.5 28.2 13.2 28.5 14.5 32 13.2 28.5 13.5 27.8 17 26.5 13.6 27.8 12.9 27.5 11.5 24" opacity="0.6"/></g></g></svg>
```

**Comment by creviera** [comment]
looks great, but just double checking that this is how it is intended to look:

![Screenshot from 2020-10-07 13-27-28](https://user-images.githubusercontent.com/4522213/95384184-18935e00-08a1-11eb-8a50-cfe2812b8da1.png)

versus

![Screenshot from 2020-10-07 13-29-05](https://user-images.githubusercontent.com/4522213/95384205-1df0a880-08a1-11eb-83f1-72f5f3149440.png)

**Comment by ninavizz** [comment]
Yup! Do you have a preference? TY for the screenshots, as always!

**Comment by ninavizz** [comment]
(oof, wnating that "see attachment" and "cannot decrypt reply" to all be italicized and with the em-dashes...

**Comment by eloquence** [comment]
"see attachment" is just a test message from a source, not a software message.

**Comment by creviera** [comment]
oh yeah those messages might be confusing too because i was using some old screenshot examples. right now "cannot decrypt reply" is italicized :heavy_check_mark: but no em-dashes.

**Comment by creviera** [comment]
@ninavizz - for now, I prefer leaving it with the original sparkles until we compare the new one with transparency some more with multicolor badges - it's not guaranteed that the badge color will be blue, so the transparency makes it harder to see a distinct sparkle shape with lighter colors.

**Comment by ninavizz** [comment]
@creviera Is this release of the PR shipping with multiple colors?? I'd thought it was only shipping with the SD Blue?

**Comment by creviera** [comment]
good point, i think i was over thinking this earlier. now's a good time to update this in the pr actually, and we can get more people testing it this way.
